### PR TITLE
[CALCITE-2648] Output collation of EnumerableWindow is not consistent…

### PR DESCRIPTION
**Pull request by zhztheplayer**
… with its implementation

JIRA link: https://issues.apache.org/jira/browse/CALCITE-2648

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java [hunk 1]
```diff
@@ -246,8 +251,17 @@ protected RelNode makeRel(RelOptCluster cluster, RelTraitSet traitSet,
...
+                = traitSet.getTraits(RelCollationTraitDef.INSTANCE);
+            RelNode window = LogicalWindow.create(cluster, relBuilder, input, program);
+            if (collations == null || collations.isEmpty() || collations.size() != 1) {
+              return window;
+            }
+            RelCollation collation = collations.get(0);
+            if (Objects.equals(collation, RelCollations.EMPTY)) {
+              return window;
+            }
+            return LogicalSort.create(window, collation, null, null);
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java]: What is the point of manually creating `LogicalSort`?

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdCollation.java [hunk 2]
```diff
@@ -275,9 +276,10 @@ private RelMdCollation() {}
    * from each of its windows. Assuming (quite reasonably) that the
    * implementation does not re-order its input rows, then any collations of its
    * input are preserved. */
+  @Deprecated // to be removed before 2.0
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdCollation.java]: Why do you mark it as deprecated?

**Comment by vlsi** [comment]
@zhztheplayer , it looks like you mark `EnumerableWindow` to return "empty collation". Would you please just return proper collation?
For instance, if no partition is present, then `EnumerableWindow` keeps `order by` collation.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java (hunk 1 above)
Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java]: If LogicalProject generates collations, then its parent rel could probably asks for a Window rel with collation. In this case a explicit sort is needed to make the window match the order.

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdCollation.java (hunk 2 above)
Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdCollation.java]: I suppose users would use this class since it provides public util methods for computing collations.

Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdCollation.java]: And the method RelMdCollation#window will always return empty collations -- this is covered by the collation method for general RelNode, so in fact this method can be removed.

**Comment by vlsi** [comment]
@zhztheplayer , I'm afraid I have no clue here, so you can disregard my comments.

I'm puzzled how even LogicalProject collation could work. Current collation for Project takes the value from its input, thus it basically forces each and every Project implementation to keep order of the rows which might easily break in case the project is performed via multithreaded algorithm.

I've no idea whether collation should be defined at LogicalWindow level, at EnumerableWindow level or for both of them.

**Comment by zhztheplayer** [comment]
I believe that current collation's implementation is basically for Calcite's enumerable rels, and also for data systems that have "deterministic" row ordering just like Calcite's enumerable rels, so myself have no objection with this design.

I am not sure if Calcite provides a way to deeply custom the collation's implementation, at least users could disable the collation traits in their planner config.

File: core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java [hunk 3]
```diff
@@ -274,12 +277,24 @@ public RexNode visitLocalRef(RexLocalRef localRef) {
       final RexInputRef ref = (RexInputRef) refToWindow.get(index);
       projectList.add(ref);
     }
-
-    return relBuilder.push(window)
+    RelNode sorted = requireSorted(window,
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: Why is this added?

Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: In the first commit of the PR the Sort has been added in WindowedAggRelSplitter, the solution is not good because it might produce Sort-Project-Window, which causes failure of PlannerTest#testDuplicateSortPlanWithOver.
The correct result should be Project-Sort-Window, so I move the change from WindowedAggRelSplitter to here.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: It still does not answer the question why do you need to add a sort node here.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: If you go with "LogicalWindow has empty collation" approach, then you should not add Sort nodes. What's the purpose? Collation is empty after all.

Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: For example, If we convert
```
LogicalProject[collation: [0]]
|-LogicalProject[collation: [0]]
```
to
```
LogicalProject[collation: [0]]
|-LogicalWindow[collation: []]
```
with no Sort node added, If the implementation of LogicalWindow has empty collation too, then the node `LogicalProject[collation: [0]]` can not be implemented because there is no input which is with a suitable collation.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: > But the converted Rel's collation should always match the original Rel's

This is false.

Rel can alter collation as it wishes.
For instance, Sort rel has different collation than its input.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: I suggest to set EMPTY collation for Project nodes that have OVER expressions.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: See `c)` in TLDR section of the comment:  https://issues.apache.org/jira/browse/CALCITE-2648?focusedCommentId=16734082&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16734082

Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: > But the converted Rel's collation should always match the original Rel's

I understand your point. 

What if adding a rule to add a Sort node on top of the LogicalWindow?

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: It must not be done

Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: > See c) in TLDR section of the comment
I understand your suggestion.

And could you please explain a bit why it is better to make window function call have no order assurance? Is that a trade-off?
If window function call could not preserve order, why a simple project should preserve order? Is there some principles of this?

Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: > Can you please clarify why don't just just update collation for LogicalProject?
I think I should have a reason to change the current ordering behavior of Calcite. There might be some downstream projects implementing there ordered window call.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: Technically speaking, I think `LogicalWindow(window=over(partition by deptno order by empid))` should have a collation of `collation=[0:CLUSTERED, 1:ASC]`. I believe, most of the implementation would sort the rows as it is prescribed by `order by`, so it makes sense to have those properties even at logical level, so the optimizer can remove redundant sorts, etc. I also think that CalcRelSplitter should create multiple LogicalWindow rels in case windows are different (e.g. if they have incompatible `partition by` clauses).

However @julianhyde believes `LogicalWindow` should have empty collation (see his comments in CALCITE-2648). I'm sure it loses lots optimization possibilities (e.g. FilterWindowTranspose), however I don't care to spend time on discussing collation for LogicalWindow as very few people care to discuss CALCITE-2648 / CALCITE-2593.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: >And could you please explain a bit why it is better to make window function call have no order assurance? 

In case the "agreed" change is to make "LogicalWindow" to always have "empty collation", then LogicalProject with OVER expressions should better have EMPTY collation as well in order to keep consistency between different relations. In other words, if we suppose that "OVER resets collation to an unknown state" (which what means empty collation for LogicalWindow), then we should specify the same for "LogicalProject with OVER expressions" (==collation of LogicalProject that have OVER should be empty)

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: > There might be some downstream projects implementing their ordered window call.

 Implementing `LogicalProjects` that have OVER expressions is a non-trivial task. It almost sure would require to have logic like CalcRelSplitter since single LogicalProject might happen to have OVER(partition by=row_number() OVER(...) ) kind of nesting, so it is not trivial to implement.

Note that regular databases (e.g. PostgreSQL) do not care to keep collations when computing OVER expressions. I don't care to read SQL standard, however I'm almost sure standard says that OVER might alter output order of rows, so my bet is standard says LogicalProject(x=count() over (order by y), input=sort(table_scan, order by=z)) might produce whatever sort order.

**Comment by zhztheplayer** [comment]
I have no cue to solve the problem perfectly, closing temporarily.

File: core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java (hunk 3 above)
Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java]: Sorry for the late reply, I've been working on some other issues these days.

>   I believe, most of the implementation would sort the rows as it is prescribed by order by, so it makes sense to have those properties even at logical level, so the optimizer can remove redundant sorts, etc.

@vlsi - Wouldn't having the properties in Enumerable level or user-defined convention level helps optimizer remove redundant sorts as well? SortRemoveRule doesn't force its input to be a logical node.

> In case the "agreed" change is to make "LogicalWindow" to always have "empty collation", then LogicalProject with OVER expressions should better have EMPTY collation as well in order to keep consistency between different relations.

Agreed. And I would write separate methods for logical window and enumerable window in RelMdCollation, returning empty collation/implementation collation respectively. Of course we should achieve agreement on whether to wipe LogicalWindow's collation before doing that.
